### ISSUE TITLE
BUG: fillna(0) does not fill NaNs

### ISSUE THREAD

**Issue by evyasonov** [pandas-dev/pandas#39926]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

---

#### Code Sample, a copy-pastable example

```python
f = pandas.DataFrame(data = [[0, 0]], columns = ['A', 'B']).astype('Int64')
f['C'] = f['A'] / f['B']
f.fillna(0)
```

Output:
```
	A	B	C
0	0	0	NaN
```

#### Problem description

fillna does not work as expected. It do not fill na

#### Expected Output
```
	A	B	C
0	0	0	0
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 7d32926db8f7541c356066dcadabf854487738de
python           : 3.7.9.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.18362
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 1.2.2
numpy            : 1.18.3
pytz             : 2020.5
dateutil         : 2.8.1
pip              : 20.3.3
setuptools       : 51.3.3.post20210118
Cython           : 0.29.21
pytest           : 6.2.1
hypothesis       : None
sphinx           : 3.4.3
blosc            : None
feather          : None
xlsxwriter       : 1.3.7
lxml.etree       : 4.6.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.19.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 0.8.3
fastparquet      : 0.3.3
gcsfs            : None
matplotlib       : 3.2.2
numexpr          : 2.7.2
odfpy            : None
openpyxl         : 3.0.6
pandas_gbq       : None
pyarrow          : 0.16.0
pyxlsb           : None
s3fs             : None
scipy            : 1.5.2
sqlalchemy       : 1.3.21
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : None
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.51.2

</details>

**Comment by phofl** [comment]
I don't think that fillna is the issue here. The vaue in C should be pd.NA not np.nan. I am not sure if fillna is supposed to work for np.nan in a Float64 Array. Maybe @jorisvandenbossche can help here?

**Comment by evyasonov** [comment]
@phofl 
The point that according to documentation I expect that fillna will fill all nans regardless of the origin. It's really confusing when it does not do its job..

**Comment by phofl** [comment]
I am not sure if this is the expected behavior here, hence I pinged joris. 

If this is the expected behavior in case of np.nan in Float64Arrays, then we should adjust the documnetation, if not then we got 2 *unrelated* bugs here.

**Comment by simonjayhawkins** [comment]
the discussion on how to interpret np.nan in a nullable float array is ongoing. #32265

so for now this is not yet a bug. but indeed as part of #32265 the documentation may need to be updated to clarify the behavior for nullable float arrays.

**Comment by phofl** [comment]
Thx, then let's repurpose this:

```
f = pandas.DataFrame(data = [[0, 0]], columns = ['A', 'B']).astype('Int64')
f['C'] = f['A'] / f['B']
```

this should return 

```
   A  B     C
0  0  0  <NA>
```

and not

```
   A  B     C
0  0  0  NaN
```

**Comment by phofl** [comment]
Could maybe someone rename the issue?

**Comment by simonjayhawkins** [comment]
I think the return of `f['C'] = f['A'] / f['B']` is correct. `A` and `B` are not missing values and therefore `C` is not missing but correctly np.NAN.

**Comment by phofl** [comment]
In this case forgot what I have said. Did not think about this.

**Comment by jorisvandenbossche** [comment]
Indeed, as @simonjayhawkins mentioned, this is still being discussed in #32265. So for now this is a bit "grey area". 

There are two different aspects being mentioned here (and I think both are still open for discussion):

- Should 0/0 in a nullable integer column result in NaN or NA? (strictly speaking it is NaN (to follow numpy), but we could also decide to deviate and make this NA instead to keep it closer to existing behaviour)
- Should NaN (if present, in addition to NA) also be interpreted as a "missing value" or not (eg return true for `isna()`, get filled with `fillna`, etc)?

But so that's discussion for #32265, which is a bit stalled at the moment, but something we need to revive.


(let's keep the issue open for now, to have a reference for this aspect if people search for it, until we make a final decision on the topic)

**Comment by jbrockmendel** [comment]
The OP is right that `f['A'] / f['B']` should have a np.nan entry, not a pd.NA entry.  If we take pd.NA seriously as meaning "missing", then it unequivocally is not missing when we do division by zero.  If we don't take it seriously, then Float64Dtype becomes pointless.

**Comment by Me-williams** [comment]
fillna(0) does seem to catch np.nan for me in pandas 1.3.3, however, if using it on a slice of a dataframe inplace doesn't work

**Comment by YunJD** [comment]
I'm not sure what to make of this, but `replace([pd.NA, np.NaN], 0)` also does not work.  In that sense I'm wondering how one actually goes about filling in values aside from changing the dtype?  Also, I get different return values using `.iloc`.  Full sample code:

```python
>>> df = pd.DataFrame({'a': [0,0], 'b': [0,0]}, dtype=pd.Int64Dtype())
>>> df
   a  b
0  0  0
1  0  0
>>> df['c'] = df.a / df.b
>>> df
   a  b    c
0  0  0  NaN
1  0  0  NaN
>>> df.replace([pd.NA, np.NaN], 0)
   a  b    c
0  0  0  NaN
1  0  0  NaN
>>> df.iloc[0].c # They print out different values?
<NA>
```

**Comment by Adam-Sparrow** [comment]
Sorry to necro the thread, I am a learner who has run into some confusion which I believe is linked to this current thread, thread https://github.com/pandas-dev/pandas/issues/42751 and https://github.com/pandas-dev/pandas/issues/25288 . Since google brought me here, I hope my questions and work arounds can help others who encounter this same issue, either using the workaround or finding the correct way to handle null_types, even if its just a minor unknown coercion method.

Throughout my code I have been using np.nan to compare columns and this raised no errors until I introduced an outer join.

Using inner join this worked:
```
diff_df = pd.merge(df_646, poc_df, left_index=True, right_index=True, suffixes=("_new", "_poc"))
cols = df_646.columns.to_list()

ordered_columns = []

for col in cols:
    ordered_columns.append(col)
    ordered_columns.append(col + '_new')
    ordered_columns.append(col + '_poc')

ordered_columns = [col for col in ordered_columns if col in diff_df.columns]

remaining_columns = sorted([col for col in diff_df.columns.to_list() if col not in ordered_columns])

all_columns = [col for col in ordered_columns]

all_columns.extend(remaining_columns)

diff_df = diff_df[all_columns]

for num in range(0, len(ordered_columns), 2):
    current_new_col = ordered_columns[num]
    current_poc_col = ordered_columns[num +1]
    comparison_col = re.sub('_new', '_comp', ordered_columns[num])
    diff_df[comparison_col] = diff_df.apply(lambda x: str(x[current_new_col]).lower() == str(x[current_poc_col]).lower(), axis=1)

    diff_df[current_poc_col] = diff_df.apply(lambda x: x[current_poc_col] if x[current_poc_col] != x[current_new_col] else np.nan, axis=1)
```

However once I introduced how='outer' to the merge function, the last line started producing a TypeError boolean value of NA is ambiguous.

To prevent this Error I tried using fillna(0) on diff_df, but I found that only the columns with the pd.NA value produced by the outer join were willed with 0. which makes my comparison invalid.

Next I tried diff_df = diff_df.replace({pd.NA: np.nan}) and fillna(np.nan), but neither method made any changes to the dataframe.

I have worked around this by using fillna(0) twice, the first operation filled the pd.NA the second operation filled the np.nan. However, I then encountered issues reconverting the values back to NULL.

I have tried diff_df = diff_df.replace({0: np.nan}) , but the values remembered their null types. Previous pd.NA fields went back to pd.NA and previous np.NaN fields went back to np.NaN.

This will work for my current task (its not a task I wil be repeating) and .info() seems to correctly count pd.NA and np.nan in the non-null counts.

However, I expect I will need to make some changes to use the final dataframe, and I now expect that going forward if I want to use methods like dropna() I will have to remember to do it Twice. So I worked at it a little more and found that this lambda function sucessfully aligned all the null types to np.nan which should alleviate my downstream concerns.
```

for col in diff_df.columns.to_list():
    diff_df[col] = diff_df[col].apply(lambda x: np.nan if str(x).isnumeric() and float(x) == 0 else x)
```

Reading the threads, it seems like this is intended behaviour, however I don't really understand the difference between null types and their interactions and I am sure there must be easier methods to produce the same results that I produced in a more pythonic way.

As I understand it, pd.NA means not found, it is not a number or a string, so I played around with the column types some more, and found that if I performed some

GENERATES  TYPERROR:

> df_646['UUID'] dtype before cleaning is Int64, all other cols are objects
> poc_df['UUID'] dtype before cleaning is Int64, all other cols are objects
> 
> ```
> 
> diff_df = pd.merge(df_646, poc_df, left_index=True, right_index=True, suffixes=("_new", "_poc"), how='outer')
> ```
> diff_df['UUID_new'] dtype after join without cleaning is Int64, all other cols are objects.
> diff_df['UUID_poc'] dtype after join without cleaning is Int64, all other cols are objects.
> the UUID dtypes have not changed.
> 

NO TYPEERROR AND NO WORKAROUND

> ```
> 
> df_646['UUID'] = df_646['UUID'].astype(int)
> poc_df['UUID'] = poc_df['UUID'].astype(int)
> ```
> df_646['UUID'] dtype after cleaning is Int64, all other cols are objects
> poc_df['UUID'] dtype after cleaning is Int64, all other cols are objects
> ```
> 
> diff_df = pd.merge(df_646, poc_df, left_index=True, right_index=True, suffixes=("_new", "_poc"), how='outer')
> ```
> diff_df['UUID_new'] dtype after join with cleaning = float64, all other cols are objects
> diff_df['UUID_poc'] dtype after join with cleaning = int64, all other cols are objects 
> df_646 UUID has changed from dtype Int64 to float64.
> poc_df UUID has changed from dtype Int64 to int64 (lower case).

I assume the errors must be due to this change of column formatting, I think forcing the UUID column to be numeric before the join prevented the appearance of pd.NA values.

It seems weird that the cleaning makes the dtype of UUID col change on join, and it also seems like that change is related to the correct processing of the values as np.nan, I would have thought that maintaining the dtype would have been the method which does not produce a failure, but I guess the null strings produced in the UUID column join are not treated numerically in that case?

Am I correct that this process is intended behaviour, and are there any useful methods that I am not aware of that could have made this process smoother? And also is my analysis of the behaviour correct?

I am using python 3.13.5 and pd__version__ 2.3.2
